PSQLADM-51 : Update mysql-server_version variable

rameshvs02 opened the pull request:
Modified `proxysql-admin` script to update mysql-server_version variables in proxysql db based on MySQL server version. `proxysql-admin` will update mysql-server_version with following options
```
proxysql-admin --enable
proxysql-admin --update-cluster
```

rameshvs02 committed:
PSQLADM-51 : Modified proxysql-admin script to update mysql-server_version variables in proxysql db based on MySQL server version

rameshvs02 committed:
PSQLADM-51 : Modified proxysql-admin script to update mysql-server_version variables in proxysql db based on MySQL server version

kennt-percona commented:
Some general comments:
(1) The two commits should be merged into one commit.
(2) How will this work with multi-cluster scenarios? 5.6/5.7 clusters using the
same proxysql instance.  (maybe it would be better to add a --update-mysql-version or --skip-mysql-version-update option, depending on what the default action is).
(3) Needs test cases
(4) Also needs doc updates

kennt-percona commented on proxysql-admin:
```diff
@@ -1507,7 +1507,20 @@ function enable_proxysql() {
                            "\n-- Please check the ProxySQL connection parameters and status."
     fi
   fi
+  
```
This should probably be a function (taking the server to use as a parameter).

rameshvs02 replied:
Added update_mysql_version function

kennt-percona commented on proxysql-admin:
```diff
@@ -1507,7 +1507,20 @@ function enable_proxysql() {
   fi
+  
+  # Update mysql server version details in proxysql db 
+  proxysql_mysql_version_string=$(proxysql_exec "$LINENO" "select variable_value from global_variables where variable_name like 'mysql-server_version'")
```
Every call to proxysql_exec and mysql_exec needs to be checked for an error return.
Brings up an interesting question, what to do if the mysql_exec call fails.  Do you want to empty the version string?

kennt-percona commented on proxysql-admin:
```diff
@@ -2063,6 +2076,20 @@ function update_cluster()
     exit 1
   fi
 
+  # Update mysql server version details in proxysql db 
```
The server used by mysql_exec may not be online at this point.  This check needs to be done further down in the code where we have determined the list of nodes.

kennt-percona commented:
Actually, it may be better to leave the default to do nothing.  Then add a --update-mysql-version option.  Also, add the relevant option to the config file, so that the default value can be set there.
This way, we don't force a behavioral change.  Also the command should output what happened ("ProxySQL MySQL version changed to x.x.x.)

rameshvs02 committed:
PSQLADM-51 : Added --update-mysql-version option to update the mysql version in proxysql db

rameshvs02 commented:
> Actually, it may be better to leave the default to do nothing. Then add a --update-mysql-version option. Also, add the relevant option to the config file, so that the default value can be set there.
> This way, we don't force a behavioral change. Also the command should output what happened ("ProxySQL MySQL version changed to x.x.x.)

Added --update-mysql-version option to update the mysql version in ProxySQL DB. update_mysql_cluster function will also print the following output. Added this option in test suite and document.

kennt-percona commented on proxysql-admin:
```diff
@@ -267,6 +267,8 @@ One of the options below must be provided.
   --enable, -e                       Auto-configure Percona XtraDB Cluster nodes into ProxySQL
   --update-cluster                   Updates the cluster membership, adds new cluster nodes
                                      to the configuration.
+  --update-mysql-version             Updates mysql server version in proxysql db based on
```
May want to say something more like
Updates the mysql-server_version variable in ProxySQL with the version from a node in the cluster.

Also applies above (to the readme.md)

rameshvs02 replied:
Changed text.

kennt-percona commented on proxysql-admin:
```diff
@@ -1021,6 +1023,45 @@ function get_all_hostgroups_from_writer_hostgroup()
+#
+# Arguments:
+#   None
+
```
Minor nitpick, should follow the formatting for the other functions.  No blank lines before the function name.

rameshvs02 replied:
Done

kennt-percona commented on proxysql-admin:
```diff
@@ -1021,6 +1023,45 @@ function get_all_hostgroups_from_writer_hostgroup()
+  CLUSTER_PORT=$(echo -e "$cluster_node" | cut -f2)
+  cluster_connection_check
+  
+  proxysql_mysql_version_string=$(proxysql_exec "$LINENO" "select variable_value from global_variables where variable_name like 'mysql-server_version'")
```
All calls to proxysql_exec() should have a check_cmd (unless you're actually testing for ProxySQL connectivity).  Similarly for mysql_exec().

rameshvs02 replied:
Done

kennt-percona commented on proxysql-admin:
```diff
@@ -3152,6 +3199,9 @@ function main() {
   elif [[ $REPORT_STATUS -eq 1 ]]; then
 
     report_status $WRITER_HOSTGROUP_ID
+  elif [[ $UPDATE_MYSQL_VERSION -eq 1 ]]; then
```
formatting, should be an empty line before the elif

rameshvs02 replied:
Done

kennt-percona commented on proxysql-admin:
```diff
@@ -3013,6 +3058,7 @@ function parse_args() {
   readonly SYNCMULTICLUSTERUSERS
   readonly QUICK_DEMO
   readonly UPDATE_CLUSTER
+  readonly UPDATE_MYSQL_VERSION
```
Comment doesn't belong here, but nowhere else to put this.  Where is the initial declaration of UPDATE_MYSQL_VERSION, plus setting it to the default value?

rameshvs02 replied:
Declared UPDATE_MYSQL_VERSION variable.

kennt-percona commented on tests/proxysql-admin-testsuite.bats:
```diff
@@ -121,13 +121,24 @@ fi
+}
+
+@test "run the check for --update-mysql-version ($WSREP_CLUSTER_NAME)" {
+  local check_mysql_version=$(mysql_exec "$HOST_IP" "$PORT_3" "SELECT VERSION();" | tail -1 | cut -d'-' -f1)
```
For the test, may want to set mysql-server_version to a known value, say "0.0.0" to be sure that we test it gets changed.

rameshvs02 replied:
Since we are checking the version between ProxySQL DB and MySQL Cluster do we really need to test with known value?

kennt-percona commented on tests/proxysql-admin-testsuite.bats:
```diff
@@ -121,13 +121,24 @@ fi
+@test "run the check for --update-mysql-version ($WSREP_CLUSTER_NAME)" {
+  local check_mysql_version=$(mysql_exec "$HOST_IP" "$PORT_3" "SELECT VERSION();" | tail -1 | cut -d'-' -f1)
+  local check_proxysql_mysql_version=$(proxysql_exec "select variable_value from global_variables where variable_name like 'mysql-server_version'" | awk '{print $0}')
+  [ "$check_proxysql_mysql_version" == "$check_mysql_version" ]
```
It's useful to output the version values before the test, so if it fails we can see what the values were (see the --version check in generic-test.bats)

rameshvs02 replied:
Done

kennt-percona commented on tests/proxysql-admin-testsuite.bats:
```diff
@@ -121,13 +121,24 @@ fi
 }
 
 
+@test "run proxysql-admin --update-mysql-version ($WSREP_CLUSTER_NAME)" {
```
It seems like this can be merged into the test below.

rameshvs02 replied:
Merged

rameshvs02 committed:
Added --update-mysql-version option to update the mysql version in proxysql db. Updated test suite and README.md accordingly.
